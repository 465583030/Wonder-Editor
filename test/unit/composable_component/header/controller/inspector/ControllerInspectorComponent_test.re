open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("controller inspector component", () => {
    let addCameraGroupInBoxWithOneComponent = () => {
      let boxComponentCount = ComponentDomTool.getBoxComponentCount() + 1;
      let cameraCategoryDomIndex =
        ComponentDomTool.getCameraCategoryDomIndex();
      let cameraGroupTypeDomIndex =
        ComponentDomTool.getCameraGroupTypeDomIndex();

      OperateComponentEventTool.addComponentIntoCurrentGameObject(
        boxComponentCount,
        cameraCategoryDomIndex,
        cameraGroupTypeDomIndex,
      );
    };
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();

      MainEditorSceneTool.initState(~sandbox, ());

      CurrentSelectSourceEditorService.setCurrentSelectSource(
        EditorType.SceneTree,
      )
      |> StateLogicService.getAndSetEditorState;

      ControllerTool.stubRequestAnimationFrame(
        createEmptyStubWithJsObjSandbox(sandbox),
      );
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
    describe("test add component", () =>
      describe("test add component in edit and run engine", () => {
        beforeEach(() =>
          MainEditorSceneTool.createDefaultScene(
            sandbox,
            MainEditorSceneTool.setFirstCameraTobeCurrentSceneTreeNode,
          )
        );
        /* TODO test add arcball if no cameraGroup */

        describe("test add arcballCameraController", () => {
          test(
            "test is run, add arcballCameraController in box, and the box haven't cameraGroup, the box shouldn't bind event",
            () => {
              ControllerTool.run();

              MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode();
              AddableComponentTool.addArcballCameraInBox();

              let editEngineState = StateLogicService.getEditEngineState();
              let runEngineState = StateLogicService.getRunEngineState();
              let currentSceneTreeNode =
                GameObjectTool.unsafeGetCurrentSceneTreeNode();

              (
                editEngineState
                |> GameObjectComponentEngineService.getArcballCameraControllerComponent(
                     StateLogicService.getEditEngineComponent(
                       DiffType.GameObject,
                       currentSceneTreeNode,
                     ),
                   )
                |. ArcballCameraEngineService.isBindArcballCameraControllerEvent(
                     editEngineState,
                   ),
                runEngineState
                |> GameObjectComponentEngineService.getArcballCameraControllerComponent(
                     currentSceneTreeNode,
                   )
                |. ArcballCameraEngineService.isBindArcballCameraControllerEvent(
                     runEngineState,
                   ),
              )
              |> expect == (false, false);
            },
          );
          test(
            "test is run, add arcballCameraController in active camera, the active camera should bind event",
            () => {
              ControllerTool.run();

              AddableComponentTool.addArcballCameraInCamera();

              let editEngineState = StateLogicService.getEditEngineState();
              let runEngineState = StateLogicService.getRunEngineState();
              let currentSceneTreeNode =
                GameObjectTool.unsafeGetCurrentSceneTreeNode();

              (
                editEngineState
                |> GameObjectComponentEngineService.getArcballCameraControllerComponent(
                     StateLogicService.getEditEngineComponent(
                       DiffType.GameObject,
                       currentSceneTreeNode,
                     ),
                   )
                |. ArcballCameraEngineService.isBindArcballCameraControllerEvent(
                     editEngineState,
                   ),
                runEngineState
                |> GameObjectComponentEngineService.getArcballCameraControllerComponent(
                     currentSceneTreeNode,
                   )
                |. ArcballCameraEngineService.isBindArcballCameraControllerEvent(
                     runEngineState,
                   ),
              )
              |> expect == (false, true);
            },
          );
        });

        describe("test add cameraGroup", () =>
          test(
            "test is run, if the currentGameObject has arcballCameraController, add cameraGroup should bind event",
            () => {
              HeaderTool.triggerAddBox();

              SceneTreeNodeDomTool.OperateDefaultScene.getNewGameObjectDomIndex()
              |> SceneTreeTool.clearCurrentGameObjectAndSetTreeSpecificGameObject;
              AddableComponentTool.addArcballCameraInBox();

              ControllerTool.run();

              addCameraGroupInBoxWithOneComponent();
              let editEngineState = StateLogicService.getEditEngineState();
              let runEngineState = StateLogicService.getRunEngineState();
              let currentSceneTreeNode =
                GameObjectTool.unsafeGetCurrentSceneTreeNode();

              (
                editEngineState
                |> GameObjectComponentEngineService.getArcballCameraControllerComponent(
                     StateLogicService.getEditEngineComponent(
                       DiffType.GameObject,
                       currentSceneTreeNode,
                     ),
                   )
                |. ArcballCameraEngineService.isBindArcballCameraControllerEvent(
                     editEngineState,
                   ),
                runEngineState
                |> GameObjectComponentEngineService.getArcballCameraControllerComponent(
                     currentSceneTreeNode,
                   )
                |. ArcballCameraEngineService.isBindArcballCameraControllerEvent(
                     runEngineState,
                   ),
              )
              |> expect == (false, true);
            },
          )
        );
      })
    );

    describe("test remove component", () => {
      describe("test remove arcballCameraController", () => {
        beforeEach(() =>
          MainEditorSceneTool.createDefaultScene(
            sandbox,
            MainEditorSceneTool.setFirstCameraTobeCurrentSceneTreeNode,
          )
        );
        test(
          "test is run, remove currentGameObject arcballCameraController, the gameObject shouldn't bind event",
          () => {
            AddableComponentTool.addArcballCameraInCamera();

            ControllerTool.run();

            SceneTreeNodeDomTool.OperateDefaultScene.getArcballCameraComponentFromCamera()
            |> OperateComponentEventTool.removeComponentFromCurrentGameObject;

            let editEngineState = StateLogicService.getEditEngineState();
            let runEngineState = StateLogicService.getRunEngineState();
            let currentSceneTreeNode =
              GameObjectTool.unsafeGetCurrentSceneTreeNode();
            (
              editEngineState
              |> GameObjectComponentEngineService.hasArcballCameraControllerComponent(
                   StateLogicService.getEditEngineComponent(
                     DiffType.GameObject,
                     currentSceneTreeNode,
                   ),
                 ),
              runEngineState
              |> GameObjectComponentEngineService.hasArcballCameraControllerComponent(
                   currentSceneTreeNode,
                 ),
            )
            |> expect == (false, false);
          },
        );
      });
      describe("test remove cameraGroup", () =>
        test(
          "test click run, remove active cameraGroup will unbind event if it has arcballCameraController, set last cameraGroup to be active, and bind event if has arcballCameraController",
          () => {
            let (camera1, camera2) =
              AddableComponentTool.getTwoAddedArcballCameraControllerCamera(
                sandbox,
              );

            ControllerTool.run();

            SceneTreeNodeDomTool.OperateDefaultScene.getCameraGroupFromCamera()
            |> OperateComponentEventTool.removeComponentFromCurrentGameObject;

            let runEngineState = StateLogicService.getRunEngineState();

            (
              runEngineState
              |> GameObjectComponentEngineService.getArcballCameraControllerComponent(
                   camera1,
                 )
              |. ArcballCameraEngineService.isBindArcballCameraControllerEvent(
                   runEngineState,
                 ),
              runEngineState
              |> GameObjectComponentEngineService.getArcballCameraControllerComponent(
                   camera2,
                 )
              |. ArcballCameraEngineService.isBindArcballCameraControllerEvent(
                   runEngineState,
                 ),
            )
            |> expect == (true, false);
          },
        )
      );
    });
  });