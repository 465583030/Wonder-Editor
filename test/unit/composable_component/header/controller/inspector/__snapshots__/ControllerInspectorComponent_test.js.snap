// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller inspector component test add component test add component in inspector test add sourceInstance component inspector should show sourceInstance component 1`] = `
<article
  className="inspector-component"
>
  <article
    className="componentBox-component"
  >
    <div
      className="header"
    >
      <div
        className="header-triangle"
        onClick={[Function]}
      >
        <span
          className="triangle-bottom"
        />
      </div>
      <div
        className="header-title"
      >
        transform
      </div>
    </div>
    <article
      className="transform-component"
    >
      <article
        className="wonder-float-input"
      >
        <span
          className="component-label"
        >
          X : 
        </span>
        <input
          className="input-component float-input"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value="0"
        />
      </article>
      <article
        className="wonder-float-input"
      >
        <span
          className="component-label"
        >
          Y : 
        </span>
        <input
          className="input-component float-input"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value="0"
        />
      </article>
      <article
        className="wonder-float-input"
      >
        <span
          className="component-label"
        >
          Z : 
        </span>
        <input
          className="input-component float-input"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value="0"
        />
      </article>
    </article>
  </article>
  <article
    className="componentBox-component"
  >
    <div
      className="header"
    >
      <div
        className="header-triangle"
        onClick={[Function]}
      >
        <span
          className="triangle-bottom"
        />
      </div>
      <div
        className="header-title"
      >
        basicMaterial
      </div>
    </div>
    <article
      className="transform-component"
    >
      <article
        className="wonder-float-input"
      >
        <span
          className="component-label"
        >
          color : 
        </span>
        <input
          className="input-component float-input"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value="#ffffff"
        />
      </article>
    </article>
  </article>
  <article
    className="componentBox-component"
  >
    <div
      className="header"
    >
      <div
        className="header-triangle"
        onClick={[Function]}
      >
        <span
          className="triangle-bottom"
        />
      </div>
      <div
        className="header-title"
      >
        sourceInstance
      </div>
      <span
        className="header-close"
      >
        x
      </span>
    </div>
    <div>
      simulate source instance
    </div>
  </article>
  <article
    className="addable-component"
  >
    <button
      disabled={true}
      onClick={[Function]}
    >
      add component
    </button>
  </article>
</article>
`;
